### PR TITLE
ScannerRegistry and Dispatch migration compatibility

### PR DESCRIPTION
**Pull request by Ramarti**
NOTE (1): Goes after
https://github.com/forta-network/forta-contracts/pull/129

NOTE (2): Many tests stay skipped until stake allocation is compatible with `NodeRunnerRegistry`

`ScannerRegistry` and `Dispatch` support using data from the new registry after migration. Non migrated `ScannerRegistry` will return `isEnabled() == false` after migration.

Added the role `NODE_RUNNER_ADMIN`.

Fix: `ScannerRegistry` keeps holding the `StakeThreshold` per chain id, since `FortaStakingParameters` need a different address per subject. `NodeRunnerRegistry` has a common threshold. 

TODO: make FortaStaking aware of stake allocation between Scanner Nodes by NodeRunner

### CODE REVIEW

**Comment by frangio** [comment]
I'm not really sure about this design for the compatibility layer... It feels like an unnecessary indirection to put the compatibility logic in the Migration contract, and inconsistent in that there could be multiple members of the `MIGRATOR_ROLE` in `ScannerRegistry` yet there is just one `controller` variable that will be consulted for the migrated data.

I would vote for removing the compatibility logic from the Migration contract, making the ScannerRegistry aware of the NodeRunnerRegistry contract instead of the Migration contract (aka the controller), and having the ScannerRegistry consult the NodeRunnerRegistry directly instead of indirectly through the Migration.

What do you think?

**Comment by frangio** [comment]
As I read more, I realize how things are connected, and I want to suggest the following changes for consideration:

- Moving the `migrationEnd` variable from the `Migration` contract into `ScannerRegistry`, which should "turn itself off" past that date, and reject any further migrations.
- Simplifying the `Migration` contract so it only has the `migrate` and `selfMigrate` methods. Its only purpose being to isolate the temporary migration logic and the minting and burning roles in the respective registries (to be removed at a later date).

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol [hunk 1]
```diff
@@ -330,57 +342,56 @@ abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgr
...
+    function setStakeThreshold(StakeThreshold calldata newStakeThreshold) external onlyRole(NODE_RUNNER_ADMIN_ROLE) {
         if (newStakeThreshold.max <= newStakeThreshold.min) revert StakeThresholdMaxLessOrEqualMin();
-        emit StakeThresholdChanged(chainId, newStakeThreshold.min, newStakeThreshold.max, newStakeThreshold.activated);
-        _stakeThresholds[chainId] = newStakeThreshold;
+        emit StakeThresholdChanged(newStakeThreshold.min, newStakeThreshold.max, newStakeThreshold.activated);
+        _stakeThreshold = newStakeThreshold;
     }
 
     /**
      * @notice Getter for StakeThreshold for the scanner with id `subject`
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/node_runners/NodeRunnerRegistryCore.sol:368]: No, there is 2 different stake thresholds at play here:

- NodeRunnerRegistry's threshold is for the node runners
- and the other one is coded in the next review, `managedSubjectThresholds`, for the managed scanner nodes (this one is chain dependent). I realized they are different in the next review, and I think we don't need those thresholds for DelegatORs and DelegatED (NodeRunners).

I may remove them if I have time.

File: contracts/components/scanners/ScannerRegistryEnable.sol [hunk 2]
```diff
@@ -46,7 +46,6 @@ abstract contract ScannerRegistryEnable is ScannerRegistryManaged {
      * @param permission the caller claims to have.
      */
     function enableScanner(uint256 scannerId, Permission permission) public virtual {
-        if (!_isStakedOverMin(scannerId)) revert StakedUnderMinimum(scannerId);
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/scanners/ScannerRegistryEnable.sol:49]: We talked about removing it, it is actually not a requirement in `NodeRunnerRegistry` and I thought it makes sense to remove it.

File: contracts/components/dispatch/Dispatch.sol [hunk 3]
```diff
@@ -249,20 +293,54 @@ contract Dispatch is BaseComponentUpgradeable {
...
+            bool registered,
+            address owner,
+            uint256 chainId,
+            string memory metadata,
+            bool operational,
+            bool disabled
+        )
+    {
+        if (_scanners_deprecated.hasMigrationEnded()) {
+            return _nodeRunners.getScannerState(address(uint160(scannerId)));
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/dispatch/Dispatch.sol:337]: What is "turn itself of"? Return false for all the registered scanners? return data from NodeRunnerRegistry only?

[the remark this answers is not in the document]
Comment (frangio) [reply contracts/components/dispatch/Dispatch.sol:337]: Automatically disable all scanners that haven't been migrated, so return false.

[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/dispatch/Dispatch.sol:337]: As we talked in the call
1- The Scanner contract already disabled all scanners
2- Moving the migrationEndTime (now sunsettingTime) to scanners, along with the adapter logic added to the migration contract

File: contracts/components/dispatch/Dispatch.sol [hunk 4]
```diff
@@ -38,76 +44,87 @@ contract Dispatch is BaseComponentUpgradeable {
...
+     */
     function scannerRegistry() public view returns (ScannerRegistry) {
-        return _scanners;
+        return _scanners_deprecated;
+    }
+
+    /**
+     * @notice Getter for ScannerRegistry.
+     * @return ScannerRegistry.
+     */
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/dispatch/Dispatch.sol:75]: deleted

File: contracts/components/scanners/ScannerRegistry.sol [hunk 5]
```diff
@@ -8,12 +8,15 @@ import "./ScannerRegistryCore.sol";
...
+import "./IScannerMigration.sol";
 
 contract ScannerRegistry is BaseComponentUpgradeable, ScannerRegistryCore, ScannerRegistryManaged, ScannerRegistryEnable, ScannerRegistryMetadata {
-    
     string public constant version = "0.1.4";
 
+    IScannerMigration private _migration;
+
     event DeregisteredScanner(uint256 scannerId);
+    event SetMigrationController(address controller);
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/scanners/ScannerRegistry.sol]: Removed IScannerMigration and related references, plus the role is now `SCANNER_2_NODE_RUNNER_MIGRATOR_ROLE`

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol [hunk 6]
```diff
@@ -283,8 +283,8 @@ abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgr
         // _isStakedOverMin already checks for disabled, but returns true in every case if stakeController is not set.
         // since isStakedOverMin() is external, we need to keep this duplicate check.
         return _scannerNodes[scanner].registered &&
-            !_scannerNodes[scanner].disabled &&
-            getStakeController().getStakeSubjectHandler(SCANNER_SUBJECT).isStakedOverMin(scannerAddressToId(scanner));
+            !_scannerNodes[scanner].disabled; // && TODO: stake
+            // getStakeController().getStakeSubjectHandler(SCANNER_SUBJECT).isStakedOverMin(scannerAddressToId(scanner));
```
Comment (frangio) [inline contracts/components/node_runners/NodeRunnerRegistryCore.sol:287]: Guessing this will be fixed in #132

**Review by frangio — APPROVED** [review]

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol (hunk 6 above)
Comment (Ramarti) [reply contracts/components/node_runners/NodeRunnerRegistryCore.sol:287]: Yep!

File: contracts/components/scanners/ScannerRegistry.sol [hunk 7]
```diff
@@ -108,37 +143,51 @@ contract ScannerRegistry is BaseComponentUpgradeable, ScannerRegistryCore, Scann
...
         emit DeregisteredScanner(scannerId);
     }
 
+    /**
+     * Declares preference for migration from ScanerRegistry to NodeRunnerRegistry. Default is yes.
+     * @param scannerId ERC721 id
+     * @param isOut true if the scanner does not want to be migrated to the NodeRunnerRegistry (and deleted)
+     */
+    function setMigrationPrefrence(uint256 scannerId, bool isOut) external onlyOwnerOf(scannerId) {
+        optingOutOfMigration[scannerId] = isOut;
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/scanners/ScannerRegistry.sol:174]: I'm currently merging this branch to #132, I'll emit the event there :)

**Pull request merged**